src/prof_recent.c

```diff
@@ -414,38 +430,111 @@ prof_recent_alloc_max_ctl_write(tsd_t *tsd, ssize_t max) {
 		prof_recent_alloc_dummy->next = node->next;
 		--prof_recent_alloc_count;
 	} while (prof_recent_alloc_count > max);
-	prof_recent_t *end = prof_recent_alloc_dummy->next;
-	assert(begin != end);
 
 	prof_recent_alloc_assert_count(tsd);
-	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	return old_max;
 
 	/*
-	 * Asynchronously handle the tctx of the to-be-deleted nodes, so that
-	 * there's no simultaneous holdings of prof_recent_alloc_mtx and
-	 * tdata->lock.  In the worst case there can be slightly extra space
-	 * overhead taken by these nodes, but the total number of nodes at any
-	 * time is bounded by (max + sum(decreases)), where "max" means the
-	 * most recent prof_recent_alloc_max and "sum(decreases)" means the
-	 * sum of the deltas of all decreases in prof_recent_alloc_max in the
-	 * past.  This (max + sum(decreases)) value is completely transparent
-	 * to and controlled by application.
+	 * The tctx of the to-be-deleted nodes as well as the nodes themselves
+	 * will be asynchronously handled by prof_recent_alloc_async_cleanup().
```

Comment:
Let's return a chain of the nodes to be deleted, instead of having the callsite figuring it out.